### PR TITLE
VRU Simulated Sensor CDAR-158

### PR DESCRIPTION
**Pull request by davidgayman**
# PR Details

## Description

The SimulatedSensor acts as a wrapper to the CARLA Semantic Lidar Sensor, capturing the data and performing post-processing to generate a list of detected objects upon request.

The SimulatedSesnsor.get_detected_objects_in_frame() function presents the primary processing sequence. Calls to internal utility classes build DetectedObject's, representing the detected vulnerable road users. High-speed LIDAR data is collected into a DataCollector class running in a separate context which self-populates via a callback.

The processing sequence performs the following stages:
1. Truth state retrieval.
2. Prefilter by distance from sensor.
3. LIDAR hitpoint retrieval.
4. Occlusion determination using a fractional hitpoint coverage threshold scaling proportionally with the sensor-to-object distance.
5. Application of a generic noise model to the results.

Processing frames are understood to consist of one full LIDAR sensor rotation, and may include vertical scan elements.

Configurability is provided to the SimulatedSensor, utility functions computing prefiltering and occlusion threshold scaling, and the noise model. Furthermore the noise model is extensible.

## Related Jira Key

[CDAR-158](https://usdot-carma.atlassian.net/browse/CDAR-158)

## Motivation and Context

This new feature is provided to enable vulnerable road user detection using a LIDAR sensor in the context of a CARLA-driven simulation. The data is intended to feed into a CARMA V2X system for data fusion with other sensors and action by CARMA platform vehicles.

## How Has This Been Tested?

Code is being unit tested, and integration testing in test/integration/simulated_sensor_open3d_lidar.py successfully collects data into the internal DataCollector and performs cache rotation.

Further testing is required before release.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [x ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

[CDAR-158]: https://usdot-carma.atlassian.net/browse/CDAR-158?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by davidgayman** [comment]
Ok ready for full review.

**Review by MishkaMN — CHANGES_REQUESTED** [review]

File: carma_utils/sensorlib/src/SemanticLidarSensor.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+                     detected_object in detected_objects])
+
+    def compute_adjusted_detection_threshold(self, range):
+        """Compute the detection threshold appropriate for the given range."""
+        dt_dr = self.__simulated_sensor_config["detection_threshold_scaling_formula"][
+            "hitpoint_detection_ratio_threshold_per_meter_change_rate"]
+        t_nominal = self.__simulated_sensor_config["detection_threshold_scaling_formula"][
+            "nominal_hitpoint_detection_ratio_threshold"]
+
+        return dt_dr * range * t_nominal
```
Comment (MishkaMN) [inline carma_utils/sensorlib/src/SemanticLidarSensor.py:208]: From the way the object's size is being represented by the horizontal and vertical fov degrees from the sensor's perspective, I no longer think we need the adjusted threshold. The threshold can stay just a fractional (0;1]

**Review by MishkaMN — CHANGES_REQUESTED** [review]

**Review by MishkaMN — DISMISSED** [review]

**Review by willjohnsonk — COMMENTED** [review]
No major comments to add, a minor one is that you may consider updating the copyright comment dates to 2023 unless there is a specific reason we're keeping 2021 that I'm unaware of

File: carma_utils/sensorlib/src/__init__.py
Comment (paulbourelly999) [inline carma_utils/sensorlib/src/__init__.py:1]: What are these files for?

File: carma_utils/sensorlib/src/collector/__init__.py
Comment (paulbourelly999) [inline carma_utils/sensorlib/src/collector/__init__.py:1]: What are these files for?

File: carma_utils/sensorlib/test/doc/SensorTestCalculations.ods
Comment (paulbourelly999) [inline carma_utils/sensorlib/test/doc/SensorTestCalculations.ods:1]: What is this file for?

File: carma_utils/sensorlib/src/SemanticLidarSensor.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        :param vertical_fov: Vertical field of view in radians.
+        :return: Expected number of hitpoints in a scan across the specified field of view.
+        """
+        num_horizontal_points_per_scan = self.__sensor.points_per_second / self.__sensor.rotation_frequency
+        horizontal_angular_resolution = self.__sensor.horizontal_fov / num_horizontal_points_per_scan
+
+        num_vertical_points_per_scan = self.__sensor.number_of_channels
+        vertical_angular_resolution = self.__sensor.vertical_fov / num_vertical_points_per_scan
+
+        return (horizontal_fov / horizontal_angular_resolution) * (vertical_fov / vertical_angular_resolution)
```
[the remark this answers is not in the document]
Comment (MishkaMN) [reply carma_utils/sensorlib/src/SemanticLidarSensor.py:270]: yeah it is probably due to performance issue, but it is a configurable parameter, so we can change it during testing. For reference, it looks like 56k points_per_second was in the example of CARLA https://carla.readthedocs.io/en/latest/ref_sensors/

**Review by willjohnsonk — APPROVED** [review]
Looks good to me, if the unit tests on your end run without issue I think we can merge

**Review by MishkaMN — APPROVED** [review]

**Pull request merged**